Sources/Promise.swift

```
@@ -234,10 +234,23 @@ public struct Promise<Value> {
      */
     public func inDispatchGroup(_ group: DispatchGroup) -> Promise<Value> {
         return Promise { fulfill in
+            var fulfilled = false;
+            let fulfillmentBarrier = DispatchQueue(label: "ThreadSafeResultsArray.queue", attributes: .concurrent)
             group.enter()
             self.call { result in
-                fulfill(result)
-                group.leave()
+                fulfillmentBarrier.sync(flags: .barrier) {
+                    // Double fulfilling a promise in a dispatch group will result
+                    // in overcalling group.leave(). This prevents that issue and sets
+                    // the promise to error out.
+                    guard !fulfilled else {
+                        let errorResult = Result<Value, Error>.failure(PromiseError.overfulfilledPromise)
+                        fulfill(errorResult)
```

In the case that you have already fulfilled, this is going to avoid calling `leave` a second time, but it will `fulfill` a second time, which breaks the promise contract (fulfill exactly once per invocation). Is it our position that, because that only happens when the wrapped promise breaks the contract, it's OK for us to do this in the name of expressing the problem safely?

---

I think the alternative here is to log a warning and take no other action.

---

I actually have that exact same question. This could cause some interesting errors where a result in an array gets modified out from under you potentially, but I'm not sure of the alternatives. 

My initial thought was to assert and then silently toss the second fulfillment in production. I can't imagine anyone is going to try to handle this error, but it could cause some weird silent failures if double fulfilment isn't caught in debug builds.

---

Also, if we have that contract for all promises, should I be catching this at the call level instead?

---

Well, we have asserted it as a _contract_ because up to now it has been the implementor's job to uphold :) That's why there was originally no safety in the library around these zip functions. PinkyPromise intends to have a small footprint.

#78 has an idea that is pretty compatible with this, which is a promise-wrapping method that adds a contract assertion, which makes it an opt-in feature for debugging.

---

Ah, I think you must have used that one as a pattern in fact. There is a lot in common here.

---

Personally, I think our priorities could be:

1. Prevent preventable crashes and our own mistakes regarding threads (such as in #75 and #80, and perhaps here re: `leave`)
2. Avoid doing extra bookkeeping work unless we are preventing crashes
3. Offer a way users can opt in to stronger checks (like #78)

I think that would mean:

1. In our `zip` functions, we do not need to add the extra layer of multiple fulfillment detection implicitly, but users can perform that wrapping on the items that they pass to `zip`.
2. In `inDispatchGroup`, we maybe should still prevent an additional call to `leave` even though we continue to call `fulfill` as many times as the wrapped promise fulfills.

---

As I think about it, I think we probably don't have to assert the double fulfill in the general case as it should be easy to track down those issues. But in the zip / zipArray cases since it can cause some very hard to track down errors because we've made the assumption that the end-user is following the contract. That sounds like a perfect reason for an assert, and to avoid the double group.leave just ignore the second fulfill.

If we want to add something like #78 that could be useful, but something about adding an optional debugging feature feels weird to me.

---

OK, are you suggesting we also skip the second fulfill here and otherwise leave the PR as is?

---

Let me think about that. We should probably have a longer conversation in slack about it.